zerver/actions/message_edit.py

```diff
@@ -757,16 +795,21 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
     #   the acting user has moved the entire topic (as visible to them).
     #
     # This rule corresponds to checking moved_all_visible_messages.
-    #
-    # We may want more complex behavior in cases where one appears to
-    # be merging topics (E.g. there are existing messages in the
-    # target topic).
     if moved_all_visible_messages:
         assert stream_being_edited is not None
-        assert topic_name is not None or new_stream is not None
+        assert target_stream is not None
+        assert target_topic is not None
 
-        user_profiles_for_visibility_policy: Dict[int, List[UserProfile]] = defaultdict(list)
+        users_to_be_notified_ids = [user["id"] for user in users_to_be_notified]
+        orig_topic_user_profiles_to_be_notified = UserProfile.objects.filter(
+            id__in=users_to_be_notified_ids
+        )
```

Comment:
Fetching the full `UserProfile` objects for every user in the stream is expensive, and ideally we'd avoid doing so -- think chat.zulip.org where you'll be fetching 25000 full user objects.

We need to change this to implement the correct algorithm without doing this fetch -- we should only need to do work in this section for users who actually have a visibility policy for either the old topic or the new, which in most cases will be <10.

Reply:
I think a possible algorithm would be to just do 2 queries to `get_users_with_user_topic_visibility_policy` and use that output to populate a dataclass with fields `{old_topic_visibility_policy, new_topic_visibility_policy, user_id}` for all users that have a visibility policy for the source topic or the final topic, and then do all the processings to execute the changes just starting from that set of users.

Reply:
> we should only need to do work in this section for users who actually have a visibility policy for either the old topic or the new

> Fetching the full UserProfile objects for every user in the stream is expensive, and ideally we'd avoid doing so

@timabbott Actually, we also need users that **don't have any visibility policies set**. For example:
| original_topic | target_topic | final_behaviour |
| ------------------ | ----------------| --------------------- |
| `INHERIT` | `UNMUTED` | `UNMUTED` |

For this case, we **need users who don't have any visibility_policy** in orig_topic and UNMUTED in target_topic.

Reply:
That's just the union of the set of users who have a policy in the original topic with the set of users with a policy in the new topic -- we know for any users that didn't show up in one of those two queries that they have INHERIT as their policy in the that topic.